sdk/servicebus/azure-messaging-servicebus/src/main/java/com/azure/messaging/servicebus/SynchronousMessageSubscriber.java

```diff
@@ -5,114 +5,220 @@
+            }
+        }
 
-        if (work.isTerminal()) {
-            logger.info("[{}] Completed. Closing Flux and cancelling subscription.", work.getId());
-            dispose();
+        if (currentWork == null) {
+            if (wip.decrementAndGet() != 0) {
+                logger.warning("There is another worker in drainLoop. But there should only be 1 worker.");
+            }
+            // If message is not delivered to downstream, we will buffer it.
+            if (!delivered) {
+                bufferMessages.add(message);
+            }
         }
     }
 
-    @Override
-    protected void hookOnComplete() {
-        logger.info("[{}] Completed. No events to listen to.", work.getId());
-        dispose();
+    /**
+     *
+     * @param requested credits for current {@link SynchronousReceiveWork}.
+     */
+    private void requestCredits(long requested) {
+        long creditToAdd = requested - (remaining.get() + bufferMessages.size());
+        if (creditToAdd > 0) {
+            remaining.addAndGet(creditToAdd);
+            subscription.request(creditToAdd);
+        }
+    }
+
+    void queueWork(SynchronousReceiveWork work) {
+
+        logger.info("[{}] Pending: {}, Scheduling receive timeout task '{}'.", work.getId(), work.getNumberOfEvents(),
+            work.getTimeout());
+        workQueue.add(work);
+        drain();
     }
 
+    private void drain() {
+        if (workQueue.size() == 0) {
+            return;
+        }
+
+        // If someone is already in this loop, then we are already clearing the queue.
+        if (!wip.compareAndSet(0, 1)) {
+            return;
+        }
+
+        // Drain queue..
+        drainQueue();
+
+    }
+
+
+    private void drainQueue() {
+        if (isTerminated()) {
+            return;
+        }
+        currentWork = workQueue.poll();
+        if (currentWork == null) {
+            return;
+        }
+        long sentFromBuffer = 0;
+        if (bufferMessages.size() > 0) {
+            // If  we already have messages in buffer, we should send it first
+
+            while (!bufferMessages.isEmpty() || sentFromBuffer < currentWork.getNumberOfEvents()) {
+                currentWork.next(bufferMessages.poll());
+                remaining.decrementAndGet();
+                ++sentFromBuffer;
+            }
+            if (sentFromBuffer == currentWork.getNumberOfEvents()) {
+                currentWork.complete();
+                logger.verbose("[{}] Sent [{}] messages from buffer.", currentWork.getId(), sentFromBuffer);
+                drainQueue();
+            }
+        }
+        // timer to complete the current in case of timeout trigger
+        timeoutOperation = getTimeoutOperation(currentWork.getTimeout());
+
+        requestCredits(currentWork.getNumberOfEvents() - sentFromBuffer);
+    }
+
+    private Disposable getTimeoutOperation(Duration timeout) {
+        return Mono.delay(timeout)
```

Comment:
This timeout operation is not tied to a work item, it's possible that between the currentWork and when this times out, the item has changed.

```java
    private Disposable getTimeoutOperation(SynchronousReceiveWork work, Duration timeout) {
        return Mono.delay(timeout).thenReturn(work)
            .subscribe(l -> {
                synchronized (currentWorkLock) {
                    if (currentWork == work) {
                        work.complete();
                    }
                }
            });
    }
```